src/algorithm/hnswlib/hnswalg.cpp

```diff
@@ -175,47 +175,66 @@ HierarchicalNSW::getDistanceByLabel(LabelType label, const void* data_point) {
 
 bool
 HierarchicalNSW::isValidLabel(LabelType label) {
-    std::unique_lock<std::mutex> lock_table(label_lookup_lock_);
+    std::shared_lock lock_table(label_lookup_lock_);
     bool is_valid = (label_lookup_.find(label) != label_lookup_.end());
     lock_table.unlock();
     return is_valid;
 }
 
+void
+HierarchicalNSW::setBatchNeigohbors(InnerIdType internal_id,
+                                    int level,
+                                    const InnerIdType* neighbors,
+                                    size_t neigbor_count) {
+    linklistsizeint* ll_cur = getLinklistAtLevel(internal_id, level);
```

Comment:
ditto, `getLinklistAtLevel` is not thread-safe

Reply:
This part is read-only, and subsequent modifications will be placed under the protection of the lock

Reply:
What happens when one thread reads while another thread modifies ?

Reply:
I have added the lock